File: src/old/edges.ts
import { is, forEach } from '@toba/node-tools'
import { Node, Tag, Way, WayType, TravelMode } from '@toba/osm-models'
import { RouteConfig } from './types'
import { allowTravelMode } from './restriction'

/** Weight (or below) indicating way is not usable */
const cannotUse = 0
/** Pattern indicating one-way enforced in direction opposite the node order */
const reverse = /^(-1|reverse)$/
/** Pattern of values indicating one-way in node order */
const forward = /^(yes|true|1)$/
/** Pattern of values indicating one-way restriction is in effect */
const isOneWay = /^(yes|true|1|-1|reverse)$/

/**
 * Weighted connections between nodes for a given mode of travel.
 */
export class Edges {
   /** Weights assigned to node-node connections based on `RouteConfig` */
   #items: Map<number, Map<number, number>>
   #travelMode: string
   #config: RouteConfig

   constructor(config: RouteConfig, travelMode: string) {
      this.#items = new Map()
      this.#travelMode = travelMode
      this.#config = config
   }

   /**
    * Throw error if any of the given nodes aren't in the graph.
    */
   ensure(...nodes: number[]) {
      forEach(nodes, id => {
         if (!this.has(id)) {
            throw new Error(`Node ${id} does not exist in the graph`)
         }
      })
   }

   /**
    * Number of edges.
    */
   get length() {
      return this.#items.size
   }

   /**
    * Add weighted edges from way and return routable nodes.
    *
    * @see https://wiki.openstreetmap.org/wiki/Key:oneway
    */
   fromWay(way: Way): Node[] {
      /** Value of one-way tag */
      let oneway = ''
      /** Weight for an edge (node connection) — higher values are preferred */
      let weight: number = cannotUse

      if (way.tags !== undefined) {
         const roadType = way.tags[Tag.RoadType]
         const railType = way.tags[Tag.RailType]
         const junction = way.tags[Tag.JunctionType]

         oneway = way.tags[Tag.OneWay] ?? ''

         if (
            is.empty(oneway) &&
            (junction == 'roundabout' ||
               junction == 'circular' ||
               roadType == WayType.Freeway)
         ) {
            // infer one-way for roundabouts and freeways
            oneway = 'yes'
         }

         if (
            this.#travelMode == TravelMode.Walk ||
            (isOneWay.test(oneway) &&
               way.tags[Tag.OneWay + ':' + this.#travelMode] == 'no')
         ) {
            // disable one-way setting for foot traffic or explicit tag
            oneway = 'no'
         }

         if (roadType !== undefined) {
            weight = this.#config.weights[roadType] ?? cannotUse
         }

         if (railType !== undefined && weight == cannotUse) {
            // TODO: is this right? How can we arbitrarily switch to rail type?
            // see if there's another way
            weight = this.#config.weights[railType] ?? cannotUse
         }

         if (
            weight <= cannotUse ||
            !allowTravelMode(way.tags, this.#config.canUse)
         ) {
            return []
         }
      }

      for (let i = 1; i < way.nodes.length; i++) {
         const n1 = way.nodes[i - 1]
         const n2 = way.nodes[i]

         // foward travel is allowed from n1 to n2
         if (!reverse.test(oneway)) this.add(n1, n2, weight)

         // reverse travel is allowed from n2 to n1
         if (!forward.test(oneway)) this.add(n2, n1, weight)
      }
      return way.nodes
   }

   /**
    * Weighted nodes (IDs) connected to the given node.
    */
   get = (node: number) => this.#items.get(node)

   /**
    * Whether `from` node exists and, optionally, if it is connected to a `to`
    * node ID.
    */
   has(from: number, to?: number) {
      const exists = this.#items.has(from)
      return exists && to !== undefined ? this.get(from)!.has(to) : exists
   }

   /**
    * Weight for the connection between `from` node and `to` node. Zero is
    * returned if the nodes aren't connected.
    */
   weight = (from: number, to: number): number =>
      this.get(from)?.get(to) ?? cannotUse

   /**
    * Add connection weight between `from` and `to` node.
    */
   add(from: Node, to: Node, weight: number) {
      let edge: Map<number, number> | undefined = this.get(from.id)

      if (edge === undefined) {
         edge = new Map()
         this.#items.set(from.id, edge)
      }
      edge.set(to.id, weight)
   }

   /**
    * Execute method for each `toNode` that `nodeID` connects to.
    */
   each(nodeID: number, fn: (weight: number, toNode: number) => void) {
      const nodes = this.get(nodeID)
      if (nodes === undefined) return
      nodes.forEach(fn)
   }

   /**
    * Map node edges to an array using given function. This may be used, for
    * example, to map edges to route options.
    */
   map<T>(nodeID: number, fn: (weight: number, toNode: number) => T): T[] {
      const nodes = this.get(nodeID)

      if (nodes === undefined) return []

      const out: T[] = []
      nodes.forEach((weight, id) => out.push(fn(weight, id)))
      return out
   }
}

File: src/edges.rs
use crate::{restriction::allow_travel_mode, EdgeWeight, RouteConfig};
use hashbrown::HashMap;
use lazy_static::*;
use osm_models::{
    tag::{self, Tagged},
    travel_by, way_type, ElementID, Node, Way,
};
use regex::Regex;

lazy_static! {
    static ref REVERSE: Regex = Regex::new("^(-1|reverse)$").unwrap();
    static ref FORWARD: Regex = Regex::new("^(yes|true|1)$").unwrap();
    static ref ONE_WAY: Regex =
        Regex::new("^(yes|true|1|-1|reverse)$").unwrap();
}

/// Weight (or below) indicating way is not usable
pub static CANNOT_USE: EdgeWeight = 0.0;
/// Weighted connection to another node
pub type Connection = HashMap<ElementID, EdgeWeight>;

pub struct Edges<'a> {
    /// Weights assigned to node-node connections based on `RouteConfig`
    items: HashMap<ElementID, Connection>,
    travel_mode: String,
    config: RouteConfig<'a>,
}

impl<'a> Edges<'a> {
    /// Number of edges
    fn len(&self) -> usize {
        self.items.len()
    }

    /// Weighted nodes (IDs) connected to the given node
    fn get(&self, node: ElementID) -> Option<&Connection> {
        self.items.get(&node)
    }

    fn get_or_insert(&mut self, node: ElementID) -> &mut Connection {
        self.items.entry(node).or_insert(HashMap::new())
    }

    fn has(&self, from: ElementID) -> bool {
        self.items.contains_key(&from)
    }

    /// Whether `from` node exists and is connected with `to` node
    fn has_connection(&self, from: ElementID, to: ElementID) -> bool {
        self.get(from).map_or(false, |m| m.contains_key(&to))
    }

    /// Weight for the connection between `from` node and `to` node.
    /// `CANNOT_USE` is returned if the nodes aren't connected.
    fn weight(&self, from: ElementID, to: ElementID) -> EdgeWeight {
        self.get(from)
            .map_or(CANNOT_USE, |m| *m.get(&to).unwrap_or(&CANNOT_USE))
    }

    /// Add connection weight between `from` and `to` node
    fn add(&mut self, from: &Node, to: &Node, weight: EdgeWeight) {
        let edge: &mut Connection = self.get_or_insert(from.id);
        edge.insert(to.id, weight);
    }

    fn ensure(&self, nodes: Vec<ElementID>) {
        for id in nodes {
            assert!(self.has(id));
        }
    }

    /// Execute method for each `toNode` that `nodeID` connects to
    fn each(
        &self,
        node: ElementID,
        cb: &dyn Fn(&EdgeWeight, &ElementID) -> (),
    ) {
        if let Some(connections) = self.get(node) {
            for (k, v) in connections {
                cb(v, k);
            }
        }
    }

    /// Map node edges to an array using given function. This may be used, for
    /// example, to map edges to route options.
    fn map<T>(
        &self,
        node: ElementID,
        cb: &dyn Fn(&EdgeWeight, &ElementID) -> T,
    ) -> Vec<T> {
        let mut out = Vec::new();

        if let Some(connections) = self.get(node) {
            for (k, v) in connections {
                out.push(cb(v, k));
            }
        }
        out
    }

    /// Add weighted edges from way and return routable nodes
    /// https://wiki.openstreetmap.org/wiki/Key:oneway
    fn from_way(&mut self, way: &Way) -> bool {
        let mut weight: EdgeWeight = CANNOT_USE;
        let mut one_way = way.get_tag(tag::ONE_WAY).unwrap_or("");
        let road_type = way.get_tag(tag::ROAD_TYPE);
        let rail_type = way.get_tag(tag::RAIL_TYPE);
        let junction = way.get_tag(tag::JUNCTION_TYPE);

        if one_way.is_empty()
            && (junction == Some("roundabout")
                || junction == Some("circular")
                || road_type == Some(way_type::FREEWAY))
        {
            // infer one-way for roundabouts and freeways
            one_way = tag::access::ALLOWED;
        }

        if self.travel_mode == travel_by::FOOT
            || (ONE_WAY.is_match(one_way)
                && way.get_tag(&format!(
                    "{}:{}",
                    tag::ONE_WAY,
                    self.travel_mode
                )) == Some(tag::access::NONE))
        {
            // disable one-way setting for foot traffic or explicit tag
            one_way = tag::access::NONE;
        }

        if let Some(t) = road_type {
            weight = *self.config.weights.get(t).unwrap_or(&CANNOT_USE);
        }

        if weight == CANNOT_USE {
            if let Some(t) = rail_type {
                // TODO: is this right? How can we arbitrarily switch to rail type?
                // see if there's another way
                weight = *self.config.weights.get(t).unwrap_or(&CANNOT_USE);
            }
        }

        if weight <= CANNOT_USE || allow_travel_mode(way, &self.config.can_use)
        {
            return false;
        }

        for (i, _) in way.nodes.iter().enumerate() {
            if i == 0 {
                continue;
            }
            let n1 = way.nodes.get(i - 1).unwrap();
            let n2 = way.nodes.get(i).unwrap();

            // foward travel is allowed from n1 to n2
            if !REVERSE.is_match(one_way) {
                self.add(n1, n2, weight);
            }
            // reverse travel is allowed from n2 to n1
            if !FORWARD.is_match(one_way) {
                self.add(n2, n1, weight);
            }
        }

        true
    }
}
